### PR TITLE
feat: Add C codes for the chapter backtracking

### PR DESCRIPTION
**Pull request by Gonglja**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/c/chapter_backtracking/permutations_i.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/**
+ * File: permutations_i.c
+ * Created Time: 2023-06-04
+ * Author: Gonglja (glj0@outlook.com)
+ */
+
+#include "../utils/common.h"
+
+/* 定义向量类型 */
+typedef struct Vector {
```
Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:10]: vector (struct variable's name is started with the lowercase letter.)

Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:10]: Could we extract the vector to the public module? Now there are many duplicated code in different files.

File: codes/c/chapter_backtracking/permutations_i.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    Vector *res = permutationsI(iNums);
+
+    // 输出结果
+    printf("输入数组 nums = ");
+    printArray(nums, sizeof(nums) / sizeof(nums[0]));
+    printf("所有排列 res = ");
+    printVectorMatrix(res);
+
+    // 释放内存
+	delVector(res);
```
Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:156]: indentation. Have you formatted the code?

File: codes/c/chapter_backtracking/permutations_i.c [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    Vector *res = newVector();
+
+    // 前序遍历
+    backtrack(iState, nums, bSelected, res);
+    return res;
+}
+
+int main() {
+
```
Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:140]: Remove this emtpy line and others as well.

File: codes/c/chapter_backtracking/permutations_i.c [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        vectorPushback(bSelected, &select[i]);
+    }
+
+    Vector *res = newVector();
+
+    // 前序遍历
+    backtrack(iState, nums, bSelected, res);
+    return res;
+}
+
```
Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:138]: /* Driver Code */

File: codes/c/chapter_backtracking/preorder_traversal_ii_compact.c [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    for (int i = 0; i < vv->size; i++) {
+        Vector *v = (Vector *)vv->data[i];
+        for (int j = 0; j < v->size; j++) {
+            TreeNode *node = (TreeNode *)v->data[j];
+            printf("%d ", node->val);
+        }
+        printf("\n");
+    }
+}
+
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_ii_compact.c:121]: /* Driver Code */

File: codes/c/chapter_backtracking/preorder_traversal_ii_compact.c [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    v->capacity = 4;
+    v->depth = 1;
+    v->data = malloc(v->capacity * sizeof(void *));
+    return v;
+}
+
+/* 析构向量 */
+void delVector(Vector *v) {
+    if (v) {
+		if (v->depth == 0) {
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_ii_compact.c:30]: indentation

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    printTree(root);
+
+    // 回溯算法
+    Vector *state = newVector();
+	Vector *choices = newVector();
+    Vector *res = newVector();
+    vectorPushback(choices, root);
+
+    // 前序遍历
+    backtrace(state, choices, res);
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:154]: backtrack

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    printf("\r\n初始化二叉树\r\n");
+    printTree(root);
+
+    // 回溯算法
+    Vector *state = newVector();
+	Vector *choices = newVector();
+    Vector *res = newVector();
+    vectorPushback(choices, root);
+
+    // 前序遍历
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:153]: abundant comment

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    backtrace(state, choices, res);
+
+    // 输出结果
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:156]: abundant comment

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    printf("输出所有根节点到节点 7 的路径，要求路径中不包含值为 3 的节点:\n");
+    printResult(res);
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:158]: I recommend directly expand the print function in main() like cpp

```cpp
/* Driver Code */
int main() {
    TreeNode *root = vecToTree(vector<int>{1, 7, 3, 4, 5, 6, 7});
    cout << "\n初始化二叉树" << endl;
    printTree(root);

    // 回溯算法
    vector<TreeNode *> state;
    vector<TreeNode *> choices = {root};
    vector<vector<TreeNode *>> res;
    backtrack(state, choices, res);

    cout << "\n输出所有根节点到节点 7 的路径，要求路径中不包含值为 3 的节点" << endl;
    for (vector<TreeNode *> &path : res) {
        vector<int> vals;
        for (TreeNode *node : path) {
            vals.push_back(node->val);
        }
        printVector(vals);
    }
}
```

File: codes/c/chapter_backtracking/preorder_traversal_iii_compact.c [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	v->capacity = 4;
+	v->depth = 1;
+	v->data = malloc(v->capacity * sizeof(void *));
+	return v;
+}
+
+/* 析构向量 */
+void delVector(Vector *v) {
+	if (v) {
+		if (v->depth == 0) {
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_compact.c:30]: indentation

File: codes/c/chapter_backtracking/preorder_traversal_iii_compact.c [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	v->size = 0;
+	v->capacity = 4;
+	v->depth = 1;
+	v->data = malloc(v->capacity * sizeof(void *));
+	return v;
+}
+
+/* 析构向量 */
+void delVector(Vector *v) {
+	if (v) {
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_compact.c:29]: indentation

File: codes/c/chapter_backtracking/preorder_traversal_iii_compact.c [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+ * Created Time: 2023-06-04
+ * Author: Gonglja (glj0@outlook.com)
+ */
+
+#include "../utils/common.h"
+
+/* 定义向量类型 */
+typedef struct Vector {
+	int size;     // 当前向量的大小
+	int capacity; // 当前向量的容量
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_iii_compact.c:12]: indentation

**Review by krahets — COMMENTED** [review]
Hi @Gonglja , please address the comments and check the code format, thx!

**Comment by krahets** [comment]
Hi @Gonglja, will you add more commits to this PR?

**Comment by Gonglja** [comment]
> Hi @Gonglja, will you add more commits to this PR?

No more for now, the remaining ones are still being worked on.

File: codes/c/chapter_backtracking/permutations_i.c [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            backtrack(state, choices, selected, res);
+            // 回退：撤销选择，恢复到之前的状态
+            *((bool *)selected->data[i]) = false;
+            vectorPopback(state);
+        }
+    }
+}
+
+// 打印二维向量中的元素
+void printVectorMatrix(vector *vv) {
```
Comment (krahets) [inline codes/c/chapter_backtracking/permutations_i.c:40]: Is this a tool function? Please extract it to the public module like `codes/cpp/utils/print_utils.hpp`

File: codes/c/chapter_backtracking/preorder_traversal_i_compact.c [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+ * Created Time: 2023-05-10
+ * Author: Gonglja (glj0@outlook.com)
+ */
+
+#include "../utils/common.h"
+
+#define MaxLength 1024
+
+/* TreeNode 动态数组简单实现 */
+struct myArray {
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_i_compact.c:12]: Can we replace myArray with the public module `vector`?

Comment (Gonglja) [reply codes/c/chapter_backtracking/preorder_traversal_i_compact.c]: Updated.

Comment (Gonglja) [reply codes/c/chapter_backtracking/preorder_traversal_i_compact.c]: > Can we replace myArray with the public module `vector`?

Updated.

**Review by krahets — APPROVED** [review]
Thanks for the PR!

**Pull request merged**